src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/CodeStyle/CodeStyleOption2`1.cs

```diff
@@ -0,0 +1,214 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Xml.Linq;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Microsoft.CodeAnalysis.CodeStyle
+{
+    internal interface ICodeStyleOption
+    {
+        XElement ToXElement();
+        object Value { get; }
+        NotificationOption2 Notification { get; }
+        ICodeStyleOption WithValue(object value);
+        ICodeStyleOption WithNotification(NotificationOption2 notification);
+    }
+
+    /// <summary>
+    /// Represents a code style option and an associated notification option.  Supports
+    /// being instantiated with T as a <see cref="bool"/> or an <c>enum type</c>.
+    /// 
+    /// CodeStyleOption also has some basic support for migration a <see cref="bool"/> option
+    /// forward to an <c>enum type</c> option.  Specifically, if a previously serialized
+    /// bool-CodeStyleOption is then deserialized into an enum-CodeStyleOption then 'false' 
+    /// values will be migrated to have the 0-value of the enum, and 'true' values will be
+    /// migrated to have the 1-value of the enum.
+    /// 
+    /// Similarly, enum-type code options will serialize out in a way that is compatible with 
+    /// hosts that expect the value to be a boolean.  Specifically, if the enum value is 0 or 1
+    /// then those values will write back as false/true.
+    /// </summary>
+    internal class CodeStyleOption2<T> : ICodeStyleOption, IEquatable<CodeStyleOption2<T>>
```

Comment:
New internal shared type, that has the core functionality from `CodeStyleOption2`.